Added documentation for @typePolicy (#586)
apollographql/apollo-ios-dev: docs/source/caching/programmatic-cache-keys.mdx

---
title: Programmatic Cache Keys
---

> In situations where [declarative cache IDs](./cache-key-resolution) don't fit your use case, you can programmatically generate cache IDs for object types in your normalized cache.

Reviewer: ```suggestion
When [declarative cache IDs](./cache-key-resolution) don't fit your use case, you can programmatically generate cache IDs for object types in your normalized cache.
```

To configure cache keys programmatically, you can alter the [`SchemaConfiguration.cacheKeyInfo(for type:object:)`](https://www.apollographql.com/docs/ios/docc/documentation/apolloapi/schemaconfiguration/cachekeyinfo(for:object:)) method to return a `CacheKeyInfo` object.

> > **Note:** Programmatic cache key usage will take precedence over declarative cache keys. If you return a `CacheKeyInfo` for an object, it will be used to compute that object's cache key and any relevant `@typePolicy` will be ignored. Returning `nil` from this function will still fall back to using declarative cache keys.

Reviewer: ```suggestion
<Note>

Programmatic cache key usage will take precedence over declarative cache keys. If you return a `CacheKeyInfo` for an object, it will be used to compute that object's cache key and any relevant `@typePolicy` will be ignored. Returning `nil` from this function will still fall back to using declarative cache keys.

</Note>
```

## `CacheKeyInfo`

> The information needed to construct a **cache key** is represented by a [`CacheKeyInfo`](https://www.apollographql.com/docs/ios/docc/documentation/apolloapi/cachekeyinfo) value. This `struct` consists of two properties you can provide to affect how a cache key is computed:

Reviewer: ```suggestion
The information needed to construct a **cache key** is represented by a [`CacheKeyInfo`](https://www.apollographql.com/docs/ios/docc/documentation/apolloapi/cachekeyinfo) value. This `struct` consists of two properties you can provide to determine how a cache key is computed:
```

1. `let uniqueKeyGroup: String?`

    An optional **group identifier** for a set of objects that should be grouped together in the normalized cache. This is used as the first component of the **cache key**.

>     > #### Important: Cache key group uniqueness
>     >
>     > All objects with the same `uniqueKeyGroup` must have unique `id`s across all types.

Reviewer: ```suggestion
    <Caution>
    
    To ensure cache key group uniqueness, all objects with the same `uniqueKeyGroup` must have unique `id`s across all types.
    
    </Caution >
```

>     To prevent cache key collisions, cache keys will always have a **group identifier** component. When the `uniqueKeyGroup` is `nil` (the default value), to [`__typename`](https://spec.graphql.org/draft/#sec-Type-Name-Introspection) of the response object is used as the **group identifier** by default.

Reviewer: ```suggestion
    To prevent cache key collisions, cache keys will always have a **group identifier** component. When the `uniqueKeyGroup` is `nil` (the default value), the [`__typename`](https://spec.graphql.org/draft/#sec-Type-Name-Introspection) of the response object is used as the **group identifier**.
```

    If multiple distinct types can be grouped together in the cache, the `CacheKeyInfo` for each `Object` should have the same `uniqueKeyGroup`.

>     > **Tip:** By grouping objects together, their **keys** in the normalized cache will have the same prefix. This allows you to search for cached objects in the same group by their cache `id`. To learn more, read about [direct cache access](./cache-transactions)

Reviewer: ```suggestion
    <Tip>
    
    By grouping objects together, their **keys** in the normalized cache will have the same prefix. This allows you to search for cached objects in the same group by their cache `id`. To learn more, read about [direct cache access](./cache-transactions)
    
    </Tip>
```

2. `let id: String`

    The unique **cache ID** representing the object. This is used as the second component of the **cache key**.

>     > #### Important: Cache ID uniqueness
>     >
>     > The ID must be deterministic and unique for all objects with the same **group identifier** (`__typename` or `uniqueKeyGroup`).
>     >
>     >That is, the key will be the same every time for a response object representing the same entity in the cache and the same key will never be used for reponse objects representing different objects that also have the same **group identifier**.

Reviewer: ```suggestion
    <Warning>
    
    To ensure cache ID uniqueness, the `id` must be deterministic and unique for all objects with the same **group identifier** (`__typename` or `uniqueKeyGroup`).
    
    That is, a response object for the same entity will always have the same key in the cache.  Different objects, even with the same group identifier, will always have distinct keys.
    
    </Caution>
```

The normalized cache constructs cache keys with the format:

`"${GroupIdentifier}:${CacheID}"`

> Given a `CacheKeyInfo`:
> ```swift
> CacheKeyInfo(id: "123", uniqueKeyGroup: "Animal")
> ```

Reviewer: ```suggestion
Given the following a `CacheKeyInfo`:
```swift showLineNumbers=false
CacheKeyInfo(id: "123", uniqueKeyGroup: "Animal")
```
```

Apollo iOS would construct a cache key of `"Animal:123"`.

## The `SchemaConfiguration` file

The `SchemaConfiguration` file is your entry point to configuring **cache keys** for the types in your schema.

When Apollo iOS [generates code for your project](../code-generation/introduction), it will generate a set of metadata types representing the GraphQL schema for your application. One of these files is named `SchemaConfiguration.swift`.

 The code generation engine creates this file if it doesn't exist yet, but never overwrites an existing `SchemaConfiguration.swift` file. This means you can edit your schema configuration without those changes being overwritten on subsequent code generation runs.

> > *Tip:* You can configure the location of the generated schema types with the [`output.schemaTypes` option in your code generation configuration](./../code-generation/codegen-configuration).

Reviewer: ```suggestion
<Tip>

You can configure the location of the generated schema types with the [`output.schemaTypes` option in your code generation configuration](./../code-generation/codegen-configuration).

</Tip>
```

## Specifying cache IDs

The `SchemaConfiguration` contains a [`cacheKeyInfo(for type:object:)`](https://www.apollographql.com/docs/ios/docc/documentation/apolloapi/schemaconfiguration/cachekeyinfo(for:object:)) function. This function provides you a JSON response `object` and the concrete `type` of the object represented by the response.

The `object` parameter provides a JSON response from either a network request or a cache hit.

The `type` parameter provides a strongly typed [`Object`](https://www.apollographql.com/docs/ios/docc/documentation/apolloapi/object) type. This is a generated metadata type representing a concrete `type` in your GraphQL schema.

To configure how cache keys are computed from a response object, you can create and return [`CacheKeyInfo`](#cachekeyinfo) values from your implementation of [`cacheKeyInfo(for:object:)`](https://www.apollographql.com/docs/ios/docc/documentation/apolloapi/schemaconfiguration/cachekeyinfo(for:object:)).

> > **Note:** When specifying **cache IDs**, make sure that you are always fetching the fields used to construct those IDs in your operations. Any response objects that don't contain the **cache ID** fields will not be able to be [merged via cache normalization](./introduction#normalizing-responses).

Reviewer: ```suggestion
<Note>

When specifying **cache IDs**, make sure that you are always fetching the fields used to construct those IDs in your operations. Any response objects that don't contain the **cache ID** fields will not be able to be [merged via cache normalization](./introduction#normalizing-responses).

</Note>
```

  static func cacheKeyInfo(for type: Object, object: ObjectData) -> CacheKeyInfo? {
    if Unions.ClassroomPets.possibleTypes.contains(type) {
      return try? CacheKeyInfo(jsonValue: object["id"])
    }

    return nil
  }
}
```

### Grouping cached objects with `uniqueKeyGroup`

If your **cache IDs** values are guaranteed to be unique across a number of different types, you may want to group them together in the cache with a common [`uniqueKeyGroup`](#cachekeyinfo).

> See [`uniqueKeyGroup`](#cachekeyinfo) for more information on grouping cached objects.

For example, if all objects that implement `interface Animal` will have unique **cache IDs**, whether they are `Dog`, `Cat`, or any other type that implements `Animal`, they can share a `uniqueKeyGroup`.

```swift title="SchemaConfiguration.swift"
public enum SchemaConfiguration: ApolloAPI.SchemaConfiguration {
  static func cacheKeyInfo(for type: Object, object: ObjectData) -> CacheKeyInfo? {
    if type.implements(Interfaces.Pet) {
      return try? CacheKeyInfo(
        jsonValue: object["id"],
        uniqueKeyGroupId: Interfaces.Pet.name
      )
    }

    return nil
  }
}
```

> ## Caveats & limitations

Reviewer: ```suggestion
## Caveats and limitations
```

Programmatic cache key resolution has a few notable quirks and limitations you should be aware of while implementing your cache key resolution function:

1. While the cache key for an object can use a field from another nested object, if the fields on the referenced object are changed in another operation, the cache key for the dependent object will not be updated. For nested objects that are not normalized with their own cache key, this will never occur, but if the nested object is an entity with its own cache key, it can be mutated independently. In that case, any other objects whose cache keys are dependent on the mutated entity will not be updated automatically. You must take care to update those entities manually with a cache mutation.

> 2. The `object` passed to this function represents data for an object in an specific operation model, not a type in your schema. This means that [aliased fields](https://spec.graphql.org/draft/#sec-Field-Alias) will be keyed on their alias name, not the name of the field on the schema type.

Reviewer: ```suggestion
2. The `object` passed to this function represents data for an object in a specific operation model, not a type in your schema. This means that [aliased fields](https://spec.graphql.org/draft/#sec-Field-Alias) will be keyed on their alias name, not the name of the field on the schema type.
```

    If multiple distinct types can be grouped together in the cache, the `CacheKeyInfo` for each `Object` should have the same `uniqueKeyGroup`.

    <Tip>

    By grouping objects together, their **keys** in the normalized cache will have the same prefix. This allows you to search for cached objects in the same group by their cache `id`. To learn more, read about [direct cache access](./cache-transactions)

    </Tip>

2. `let id: String`

    The unique **cache ID** representing the object. This is used as the second component of the **cache key**.

    <Caution>

    To ensure cache ID uniqueness, the `id` must be deterministic and unique for all objects with the same **group identifier** (`__typename` or `uniqueKeyGroup`).

    That is, a response object for the same entity will always have the same key in the cache.  Different objects, even with the same group identifier, will always have distinct keys.

    </Caution>

The normalized cache constructs cache keys with the format:

`"${GroupIdentifier}:${CacheID}"`

Given the following a `CacheKeyInfo`:

> ```swift showLineNumbers=false
> CacheKeyInfo(id: "123", uniqueKeyGroup: "Animal")

Reviewer: Looks like the formatting is off here:

```suggestion

```swift showLineNumbers=false
CacheKeyInfo(id: "123", uniqueKeyGroup: "Animal")
```

```

> ```swift showLineNumbers=false
> CacheKeyInfo(id: "123", uniqueKeyGroup: "Animal")
> Apollo iOS would construct a cache key of `"Animal:123"`.
> 

Reviewer: Looks like the formatting is off here, high probability it was from a previous suggestion of mine 😅

```suggestion
```swift showLineNumbers=false

CacheKeyInfo(id: "123", uniqueKeyGroup: "Animal")
```

Apollo iOS would construct a cache key of `"Animal:123"`.

```
